File: wasm/examples/diff_tags.mjs
// Demonstrates the tag-diffing API via audex-wasm.
//
// Usage: node diff_tags.mjs <file_a> <file_b>
//
// Mirrors audex/examples/diff_tags.rs — compares the metadata of two audio
// files and prints the results using every available diff method.

import { readFileSync } from "node:fs";
import { resolve, basename, dirname } from "node:path";
import { fileURLToPath, pathToFileURL } from "node:url";

const __dirname = dirname(fileURLToPath(import.meta.url));
const PKG_DIR = resolve(__dirname, "../pkg");

// Load and initialise the WASM module
const wasmBytes = readFileSync(resolve(PKG_DIR, "audex_wasm_bg.wasm"));
const { initSync, AudioFile } = await import(
  pathToFileURL(resolve(PKG_DIR, "audex_wasm.js")).href
);
initSync({ module: wasmBytes });

// ── Main ────────────────────────────────────────────────────────────────

const args = process.argv.slice(2);
if (args.length < 2) {
  console.error("Usage: node diff_tags.mjs <file_a> <file_b>");
  process.exit(1);
}

const pathA = resolve(args[0]);
const pathB = resolve(args[1]);

// Load both files
const fileA = new AudioFile(new Uint8Array(readFileSync(pathA)), basename(pathA));
const fileB = new AudioFile(new Uint8Array(readFileSync(pathB)), basename(pathB));

console.log(`File A: ${pathA} (${fileA.formatName()})`);
console.log(`File B: ${pathB} (${fileB.formatName()})`);

// --- Basic diff ---
// diffTags() compares format-specific keys as-is (e.g. TIT2 vs title).
console.log("\n=== Basic diff ===");
const d = fileA.diffTags(fileB);
if (d.isIdentical()) {
  console.log("Tags are identical.");
} else {
  console.log(d.pprint());
}

// --- Summary ---
console.log("\n=== Summary ===");
console.log(d.summary());

// --- Pretty-print (right-aligned keys) ---
console.log("\n=== Pretty-print ===");
console.log(d.pprint());

// --- Diff with options (stream info + case-insensitive) ---
console.log("\n=== With options (stream info + case-insensitive) ===");
const d2 = fileA.diffTagsWithOptions(
  fileB,
  true,   // compare_stream_info
  true,   // case_insensitive_keys
  null,   // trim_values
  true,   // include_unchanged
);
console.log(d2.pprint());

// --- Full pretty-print (including unchanged) ---
console.log("\n=== Full pretty-print ===");
console.log(d2.pprintFull());

// --- Snapshot-based diffing ---
console.log("\n=== Snapshot-based diff ===");
const snapshot = fileA.snapshotTags();
const d3 = fileB.diffAgainstSnapshot(snapshot);
console.log(d3.summary());

// --- Filtering ---
console.log("\n=== Filtered diff (artist only) ===");
const filtered = d.filterKeys(["artist", "ARTIST", "TPE1", "\u00a9ART"]);
if (filtered.diffCount() === 0) {
  console.log("No artist differences.");
} else {
  console.log(filtered.pprint());
}

// --- Normalised diff ---
// diffTagsNormalized() maps keys through TagMap so that format-specific
// names become their standard equivalents (e.g. TIT2 → Title).
console.log("\n=== Normalised diff ===");
const d4 = fileA.diffTagsNormalized(fileB);
console.log(d4.summary());

// --- Normalised diff with stream info ---
console.log("\n=== Normalised diff (with stream info) ===");
const d5 = fileA.diffTagsNormalized(fileB, true, true);
console.log(d5.pprintFull());

// --- toJson() for raw data access ---
console.log("\n=== toJson() ===");
const json = d4.toJson();
console.log(`  changed: ${json.changed.length}, left_only: ${json.left_only.length}, right_only: ${json.right_only.length}, unchanged: ${json.unchanged.length}`);

// --- differingKeys() ---
console.log("\n=== Differing keys ===");
const keys = d4.differingKeys();
console.log(`  ${keys.length} keys differ: ${keys.slice(0, 10).join(", ")}${keys.length > 10 ? "..." : ""}`);

// Free all WASM objects
filtered.free();
d.free();
d2.free();
d3.free();
d4.free();
d5.free();
fileA.free();
fileB.free();

File: examples/diff_tags.rs
//! Demonstrates the tag-diffing API.
//!
//! Usage:
//!   cargo run --example diff_tags -- <file_a> <file_b>
//!
//! Compares the metadata of two audio files and prints the results
//! in several formats.

use audex::File;
use audex::diff::{self, DiffOptions};

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args: Vec<String> = std::env::args().collect();
    if args.len() < 3 {
        eprintln!("Usage: {} <file_a> <file_b>", args[0]);
        std::process::exit(1);
    }

    let file_a = File::load(&args[1])?;
    let file_b = File::load(&args[2])?;

    // --- Basic diff ---
    println!("=== Basic diff ===");
    let d = diff::diff(&file_a, &file_b);
    if d.is_identical() {
        println!("Tags are identical.");
    } else {
        println!("{}", d);
    }

    // --- Summary ---
    println!("\n=== Summary ===");
    println!("{}", d.summary());

    // --- Pretty-print (right-aligned keys) ---
    println!("\n=== Pretty-print ===");
    print!("{}", d.pprint());

    // --- Diff with options (stream info + case-insensitive) ---
    println!("\n=== With options (stream info + case-insensitive) ===");
    let opts = DiffOptions {
        compare_stream_info: true,
        case_insensitive_keys: true,
        include_unchanged: true,
        ..Default::default()
    };
    let d2 = diff::diff_with_options(&file_a, &file_b, &opts);
    println!("{}", d2);

    // --- Full pretty-print (including unchanged) ---
    println!("\n=== Full pretty-print ===");
    print!("{}", d2.pprint_full());

    // --- Snapshot-based diffing ---
    println!("\n=== Snapshot-based diff ===");
    let snapshot = diff::snapshot_tags(&file_a);
    let d3 = diff::diff_against_snapshot(&file_b, &snapshot);
    println!("{}", d3.summary());

    // --- Filtering ---
    println!("\n=== Filtered diff (artist only) ===");
    let filtered = d.filter_keys(&["artist", "ARTIST", "TPE1", "\u{00a9}ART"]);
    if filtered.diff_count() == 0 {
        println!("No artist differences.");
    } else {
        println!("{}", filtered);
    }

    // --- Convenience methods ---
    println!("\n=== Convenience method: file_a.diff_tags(&file_b) ===");
    let d4 = file_a.diff_tags(&file_b);
    println!("{}", d4.summary());

    Ok(())
}
